### PR TITLE
Replace RabbitMQ transaction with ack, reject model and refactor

### PR DESCRIPTION
**Pull request by Vathsan**
Fixes https://github.com/wso2/product-ei/issues/4835
This PR removes the transaction model in RabbitMQ and uses basicAck & basicReject model to provide Guaranteed message delivery.

Earlier, according to our transaction model implementation, we either commit or rollback the message during a transaction. This PR implements a new rabbitMQ parameter (SET_REQUEUE_ON_ROLLBACK), which can be set to either true or false (by default true). If set to true, we reject and requeue the message to the same queue upon failure when we do rollback. 

The validation of injecting sequence being null (at line 136) is moved to the initialization as part of the refactoring. This change will be backward incompatible for the users who have deployed injecting sequence after the inbound endpoint. To deploy the inbound endpoint properly, the injecting sequence should be deployed beforehand.

### CODE REVIEW

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/rabbitmq/RabbitMQInjectHandler.java [hunk 1]
```diff
@@ -127,7 +129,7 @@ public boolean invoke(RabbitMQMessage message, String inboundName) {
 
         if (injectingSeq == null || injectingSeq.equals("")) {
```
Comment (sdkottegoda) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/rabbitmq/RabbitMQInjectHandler.java]: This check shouldn't be there inside the invoke method. The initialization itself should fail in this case. WDYT?

Comment (sdkottegoda) [reply components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/rabbitmq/RabbitMQInjectHandler.java]: Let's move the validation of seq being null at line 136 to the initialization as well. This change will be backward incompatible for the users who have deployed injecting sequence after the inbound ep. So, let's clearly mention this in the description.

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/rabbitmq/RabbitMQConnectionConsumer.java [hunk 2]
```diff
@@ -149,66 +149,54 @@ private void startConsumer() throws ShutdownSignalException, IOException {
...
             if (message != null) {
                 idle = false;
+                RabbitMQAckStates ackState = RabbitMQAckStates.REJECT_AND_REQUEUE;
+                boolean mediationError = false;
                 try {
-                    successful = injectHandler.invoke(message, inboundName);
+                    ackState = injectHandler.invokeAndReturnAckState(message, inboundName);
                 } catch (Exception e) {         //we need to handle any exception upon injecting to mediation
-                    successful = false;
+                    ackState = RabbitMQAckStates.REJECT_AND_REQUEUE;
```
Comment (sdkottegoda) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/rabbitmq/RabbitMQConnectionConsumer.java]: We need to evaluate the value of the property, 'SET_REQUEUE_ON_ROLLBACK'. WDYT?

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/rabbitmq/RabbitMQInjectHandler.java (hunk 1 above)
Comment (Vathsan) [reply components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/rabbitmq/RabbitMQInjectHandler.java]: Done in 4c1a1ffdbc47a41a3590bc017bb57689210f029d

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/rabbitmq/RabbitMQConnectionConsumer.java (hunk 2 above)
Comment (Vathsan) [reply components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/rabbitmq/RabbitMQConnectionConsumer.java]: Done in db57000c3482bf048fdfba088070458e5de91930

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/rabbitmq/RabbitMQConnectionConsumer.java [hunk 3]
```diff
@@ -149,66 +149,53 @@ private void startConsumer() throws ShutdownSignalException, IOException {
...
-            boolean successful = false;
-            boolean mediationError = false;
-
-            RabbitMQMessage message = null;
-            try {
                 message = getConsumerDelivery(queueingConsumer);
-            } catch (InterruptedException e) {
-                log.error("Error while consuming message", e);
+            } catch (InterruptedException | ShutdownSignalException | ConsumerCancelledException e) {
+                log.debug("Exception occurred while consuming the message", e);
```
Comment (sdkottegoda) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/rabbitmq/RabbitMQConnectionConsumer.java]: Why debug?

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/rabbitmq/RabbitMQInjectHandler.java [hunk 4]
```diff
@@ -47,10 +49,20 @@
...
-                                 SynapseEnvironment synapseEnvironment) {
+                                 SynapseEnvironment synapseEnvironment) throws SynapseException {
         this.injectingSeq = injectingSeq;
+        if (injectingSeq == null || injectingSeq.equals("")) {
+            String msg = "Injecting Sequence name is not specified.";
+            log.error(msg);
+            throw new SynapseException(msg);
+        }
+        seq = (SequenceMediator) synapseEnvironment.getSynapseConfiguration().getSequence(injectingSeq);
+        if (seq != null && !seq.isInitialized()) {
```
Comment (sdkottegoda) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/rabbitmq/RabbitMQInjectHandler.java]: Let's throw a SynapseExcpetion if the sequence is null

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/rabbitmq/RabbitMQConnectionConsumer.java [hunk 5]
```diff
@@ -412,5 +390,20 @@ private void handleException(String msg, Exception e) {
...
+
+enum RabbitMQAckStates {
+    ACK,
+    REJECT,
+    REJECT_AND_REQUEUE
+}
 
-}
\ No newline at end of file
+class AckStatesAndMediationError {
```
Comment (sdkottegoda) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/rabbitmq/RabbitMQConnectionConsumer.java]: Would just State be a better name? It'll have the fields, state.acknowledge and state.mediationError. WDYT?

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/rabbitmq/RabbitMQConnectionConsumer.java (hunk 3 above)
Comment (Vathsan) [reply components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/rabbitmq/RabbitMQConnectionConsumer.java]: Done in d926aaf94be8b06f295411ec90556f84d278cf52

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/rabbitmq/RabbitMQInjectHandler.java (hunk 4 above)
Comment (Vathsan) [reply components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/rabbitmq/RabbitMQInjectHandler.java]: Fixed in d926aaf94be8b06f295411ec90556f84d278cf52

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/rabbitmq/RabbitMQConnectionConsumer.java (hunk 5 above)
Comment (Vathsan) [reply components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/rabbitmq/RabbitMQConnectionConsumer.java]: Done in d926aaf94be8b06f295411ec90556f84d278cf52

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/rabbitmq/RabbitMQInjectHandler.java [hunk 6]
```diff
@@ -47,10 +48,23 @@
...
+                                 SynapseEnvironment synapseEnvironment) throws SynapseException {
         this.injectingSeq = injectingSeq;
+        if (injectingSeq == null || injectingSeq.equals("")) {
+            String msg = "Injecting Sequence name is not specified.";
+            log.error(msg);
+            throw new SynapseException(msg);
+        }
+        seq = (SequenceMediator) synapseEnvironment.getSynapseConfiguration().getSequence(injectingSeq);
+        if (seq == null) {
+            throw new SynapseException("Sequence Mediator is null.");
```
Comment (sdkottegoda) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/rabbitmq/RabbitMQInjectHandler.java]: The log should be "specified injecting sequence: <name>" is invalid.

Comment (Vathsan) [reply components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/rabbitmq/RabbitMQInjectHandler.java]: Done in b459fdfdf15ed24137e23f5b3a394f2683be6223

**Pull request merged**